File: Embassat/View/Artist Detail/CADEMArtistDetailViewModel.m
//
//  CADEMArtistDetailViewModel.m
//  Embassat
//
//  Created by Joan Romano on 26/03/14.
//  Copyright (c) 2014 Crows And Dogs. All rights reserved.
//

#import "CADEMArtistDetailViewModel.h"

#import "NSString+EMAdditions.h"
#import <EventKit/EventKit.h>

@interface CADEMArtistDetailViewModel ()

@property (nonatomic, strong) CADEMArtistSwift *model;

@end

@implementation CADEMArtistDetailViewModel

#pragma mark Lifecycle

- (instancetype)init
{
	return [self initWithModel:nil];
}

- (instancetype)initWithModel:(id)model
{
	if (self = [super init])
    {
        _model = model;
        
        RAC(self, artistName) = RACObserve(self.model, name);
        RAC(self, artistDescription) = [RACObserve(self.model, longDescription) map:^id(NSString *longDescription) {
            NSDictionary *options = @{ NSDocumentTypeDocumentAttribute : NSHTMLTextDocumentType,
                                       NSCharacterEncodingDocumentAttribute :@(NSUTF8StringEncoding) };
            NSData *data = [longDescription dataUsingEncoding:NSUTF8StringEncoding];

            return [[[NSAttributedString alloc] initWithData:data options:options documentAttributes:nil error:nil] string];
        }];
//        RAC(self, artistStage) = [RACObserve(self.model, stage) map:^id(NSString *stage) {
//            return [stage isEqualToString:@"Amfiteatre Yeearphone"] ? @"AMFITEATRE" : [stage uppercaseString];
//        }];
        RAC(self, artistImageURL) = RACObserve(self.model, imageURL);
//        RAC(self, artistStartHour) = RACObserve(self.model, initialHour);
//        RAC(self, artistStartMinute) = RACObserve(self.model, initialMinute);
//        RAC(self, artistDay) = [RACObserve(self.model, date) map:^id(NSString *dateString) {
//            return [dateString isEqualToString:@"25/06/2014"] ? @"DIM" : [dateString isEqualToString:@"26/06/2014"] ? @"DIJ" : [dateString isEqualToString:@"27/06/2014"] ? @"DIV" : @"DIS";
//        }];
    }
    
	return self;
}

#pragma mark - Public

- (void)addEventOnCalendar
{
    EKEventStore *store = [[EKEventStore alloc] init];
    [store requestAccessToEntityType:EKEntityTypeEvent completion:^(BOOL granted, NSError *error) {
        if (!granted) { return; }
        EKEvent *event = [EKEvent eventWithEventStore:store];
        event.title = [NSString stringWithFormat:@"%@ @ %@", self.artistName, self.artistStage];
//        event.startDate = self.model.initialDate;
//        event.endDate = self.model.finalDate;
        [event setCalendar:[store defaultCalendarForNewEvents]];
        NSError *err = nil;
        BOOL success = [store saveEvent:event span:EKSpanThisEvent commit:YES error:&err];
        NSString *savedEventId = event.eventIdentifier;  //this is so you can access this event later
        
        if (success)
        {
            DDLogInfo(@"Event added with id: %@", savedEventId);
            dispatch_async(dispatch_get_main_queue(), ^{
                [[[UIAlertView alloc] initWithTitle:@"Embassa't" message:@"Afegit al calenari correctament" delegate:nil cancelButtonTitle:nil otherButtonTitles:@"D'acord", nil] show];
            });
        }
    }];
}

@end

File: Embassat/View/Artist Detail/CADEMArtistDetailViewModel.swift
//
//  CADEMArtistDetailViewModel.swift
//  Embassa't
//
//  Created by Joan Romano on 17/05/15.
//  Copyright (c) 2015 Crows And Dogs. All rights reserved.
//

import Foundation
import EventKit

public class CADEMArtistDetailViewModel: NSObject {
    
    var model: Array<CADEMArtist>
    let service: CADEMArtistService = CADEMArtistService()
    
    var currentArtist: CADEMArtist
    var artistName: String = ""
    var artistDescription: String = ""
    var artistStartTimeString: String = ""
    var artistDay: String = ""
    var artistStage: String = ""
    var artistIsFavorite: Bool = false
    var artistImageURL: NSURL? = nil
    
    public var currentIndex: Int = 0 {
        didSet {
            if currentIndex > count(model) - 1 {
               currentIndex = count(model) - 1
            }
            
            if currentIndex < 0 {
                currentIndex = 0
            }
            
            currentArtist = model[currentIndex]
            self.updateCurrentArtistData()
        }
    }
    
    init(model: Array<CADEMArtist>, currentIndex: Int) {
        self.model = model
        self.currentIndex = currentIndex
        
        currentArtist = model[currentIndex]
        
        super.init()
        
        self.updateCurrentArtistData()
    }
    
    public func shareAction(forViewController viewController: UIViewController) {
        let item: AnyObject! = SHKItem.URL(currentArtist.artistURL, title: String(format: "%@ @ Embassa't", self.artistName), contentType: SHKURLContentTypeUndefined)
        let actionSheet = SHKActionSheet(forItem: item as! SHKItem)
        SHK.setRootViewController(viewController)
        
        actionSheet.showFromToolbar(viewController.navigationController?.toolbar)
    }
    
    public func toggleFavorite(completion: () -> ()) {
        service.toggleFavorite(forArtist: currentArtist).subscribeNext {
            [unowned self] (updatedArtist: AnyObject!) -> Void in
            self.model[self.currentIndex] = updatedArtist as! CADEMArtist
            completion()
        }
    }
    
    func updateCurrentArtistData() {
        artistName = currentArtist.name
        artistDescription = currentArtist.longDescription
        artistStage = currentArtist.stage
        artistImageURL = currentArtist.imageURL
        artistStartTimeString = String(format: "%@:%@", String(currentArtist.startDate.hourString), String(currentArtist.startDate.minuteString))
        artistIsFavorite = currentArtist.favorite
        artistDay = currentArtist.scheduleDayString
    }
   
}
